htlcswitch/link.go

```diff
@@ -410,40 +421,52 @@ func (l *channelLink) handleDownStreamPkt(pkt *htlcPacket) {
 		// commitment chains.
 		htlc.ChanID = l.ChanID()
 		index, err := l.channel.AddHTLC(htlc)
-		if err == lnwallet.ErrMaxHTLCNumber {
-			log.Infof("Downstream htlc add update with "+
-				"payment hash(%x) have been added to "+
-				"reprocessing queue, batch: %v",
-				htlc.PaymentHash[:],
-				l.batchCounter)
-			l.overflowQueue.consume(pkt)
-			return
-		} else if err != nil {
-			failPacket := newFailPacket(
-				l.ShortChanID(),
-				&lnwire.UpdateFailHTLC{
-					Reason: []byte{byte(0)},
-				},
-				htlc.PaymentHash,
-				htlc.Amount,
-			)
-
-			// The HTLC was unable to be added to the state
-			// machine, as a result, we'll signal the switch to
-			// cancel the pending payment.
-			go l.cfg.Switch.forward(failPacket)
-
-			log.Errorf("unable to handle downstream add HTLC: %v",
-				err)
-			return
+		if err != nil {
+			switch err {
+			case lnwallet.ErrMaxHTLCNumber:
+				log.Infof("Downstream htlc add update with "+
+					"payment hash(%x) have been added to "+
+					"reprocessing queue, batch: %v",
+					htlc.PaymentHash[:],
+					l.batchCounter)
+				l.overflowQueue.consume(pkt)
+				return
+
+			default:
+				// The HTLC was unable to be added to the state
+				// machine, as a result, we'll signal the switch to
+				// cancel the pending payment.
+				onionReader := bytes.NewReader(htlc.OnionBlob[:])
+				obfuscator, err := l.cfg.DecodeOnionObfuscator(onionReader)
+				if err != nil {
+					l.sendMalformedHTLCError(err, htlc.OnionBlob[:], htlc.ID)
+					return
+				}
+
+				failure := lnwire.NewTemporaryChannelFailure(nil)
+				reason, err := obfuscator.InitialObfuscate(failure)
+				if err != nil {
+					log.Errorf("unable to obfuscate error: %v", err)
+					return
+				}
+
+				go l.cfg.Switch.forward(newFailPacket(
+					l.ShortChanID(),
+					&lnwire.UpdateFailHTLC{
+						Reason: reason,
+					}, htlc.PaymentHash, htlc.Amount, false,
```

Comment:
Shouldn't the last parameter here be set to true? As the error has already been encrypted once.